File: python_agents/utils/pii.py
"""
Comprehensive PII redaction for logs and tickets.

Masks:
- Email addresses
- Phone numbers
- Social Security Numbers (SSN)
- Credit card numbers
- Street addresses
- IP addresses (except localhost)
"""

import re
from dataclasses import dataclass
from typing import Any


EMAIL_PATTERN = re.compile(
    r'[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}'
)

PHONE_PATTERN = re.compile(
    r'(?:\+?1[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}'
)

SSN_PATTERN = re.compile(
    r'\b\d{3}[-.\s]?\d{2}[-.\s]?\d{4}\b'
)

CREDIT_CARD_PATTERN = re.compile(
    r'\b(?:\d{4}[-.\s]?){3}\d{4}\b'
)

ADDRESS_PATTERN = re.compile(
    r'\b\d{1,5}\s+[A-Za-z0-9\s]+(?:Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Drive|Dr|Lane|Ln|Court|Ct|Way|Circle|Cir)\b',
    re.IGNORECASE
)

IP_ADDRESS_PATTERN = re.compile(
    r'\b(?:\d{1,3}\.){3}\d{1,3}\b'
)

LOCALHOST_IPS = frozenset({"127.0.0.1", "0.0.0.0"})


@dataclass
class RedactorConfig:
    """Configuration for PII redaction."""
    mask_char: str = "*"
    preserve_length: bool = False
    redact_emails: bool = True
    redact_phones: bool = True
    redact_ssn: bool = True
    redact_credit_cards: bool = True
    redact_addresses: bool = True
    redact_ips: bool = True


def _mask_email(email: str, mask_char: str, preserve_length: bool) -> str:
    """Mask an email address."""
    if preserve_length:
        at_idx = email.find("@")
        if at_idx > 0:
            local = email[:at_idx]
            domain = email[at_idx + 1:]
            masked_local = local[0] + mask_char * (len(local) - 1)
            dot_idx = domain.rfind(".")
            if dot_idx > 0:
                domain_name = mask_char * dot_idx
                tld = domain[dot_idx:]
                return f"{masked_local}@{domain_name}{tld}"
            return f"{masked_local}@{mask_char * len(domain)}"
    return "[EMAIL]"


def _mask_phone(phone: str, mask_char: str, preserve_length: bool) -> str:
    """Mask a phone number."""
    if preserve_length:
        return re.sub(r'\d', mask_char, phone)
    return "[PHONE]"


def _mask_generic(value: str, mask_char: str, preserve_length: bool, label: str) -> str:
    """Generic masking for SSN, CC, addresses, IPs."""
    if preserve_length:
        return mask_char * len(value)
    return f"[{label}]"


def redact(text: str, config: RedactorConfig | None = None) -> str:
    """
    Redact PII from a string.
    
    Args:
        text: The string to redact
        config: Optional redaction configuration
        
    Returns:
        String with PII masked
    """
    if not isinstance(text, str):
        return text
    
    cfg = config or RedactorConfig()
    result = text
    
    if cfg.redact_emails:
        result = EMAIL_PATTERN.sub(
            lambda m: _mask_email(m.group(), cfg.mask_char, cfg.preserve_length),
            result
        )
    
    if cfg.redact_credit_cards:
        result = CREDIT_CARD_PATTERN.sub(
            lambda m: _mask_generic(m.group(), cfg.mask_char, cfg.preserve_length, "CC"),
            result
        )
    
    if cfg.redact_ssn:
        result = SSN_PATTERN.sub(
            lambda m: _mask_generic(m.group(), cfg.mask_char, cfg.preserve_length, "SSN"),
            result
        )
    
    if cfg.redact_phones:
        result = PHONE_PATTERN.sub(
            lambda m: _mask_phone(m.group(), cfg.mask_char, cfg.preserve_length),
            result
        )
    
    if cfg.redact_addresses:
        result = ADDRESS_PATTERN.sub(
            lambda m: _mask_generic(m.group(), cfg.mask_char, cfg.preserve_length, "ADDRESS"),
            result
        )
    
    if cfg.redact_ips:
        def mask_ip(match: re.Match) -> str:
            ip = match.group()
            if ip in LOCALHOST_IPS:
                return ip
            return _mask_generic(ip, cfg.mask_char, cfg.preserve_length, "IP")
        result = IP_ADDRESS_PATTERN.sub(mask_ip, result)
    
    return result


def redact_object(obj: Any, config: RedactorConfig | None = None) -> Any:
    """
    Recursively redact PII from any JSON-serializable object.
    
    Args:
        obj: Any JSON-serializable value (str, dict, list, etc.)
        config: Optional redaction configuration
        
    Returns:
        Object with all PII masked
    """
    if obj is None:
        return obj
    
    if isinstance(obj, str):
        return redact(obj, config)
    
    if isinstance(obj, dict):
        return {k: redact_object(v, config) for k, v in obj.items()}
    
    if isinstance(obj, (list, tuple)):
        return type(obj)(redact_object(item, config) for item in obj)
    
    return obj


class PIIRedactor:
    """
    Stateful PII redactor with configurable settings.
    
    Example:
        redactor = PIIRedactor(preserve_length=True)
        clean_data = redactor.redact_all(sensitive_data)
    """
    
    def __init__(self, **config_kwargs):
        """Initialize with optional config overrides."""
        self.config = RedactorConfig(**config_kwargs)
    
    def redact_string(self, text: str) -> str:
        """Redact PII from a string."""
        return redact(text, self.config)
    
    def redact_all(self, obj: Any) -> Any:
        """Redact PII from any object recursively."""
        return redact_object(obj, self.config)

File: lib/pii.ts
/**
 * PII Redactor
 * 
 * Masks sensitive personal information (emails, phone numbers, addresses)
 * in logs and diagnostic tickets.
 */

const EMAIL_REGEX = /[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}/g;

const PHONE_REGEX = /(?:\+?1[-.\s]?)?(?:\(?\d{3}\)?[-.\s]?)?\d{3}[-.\s]?\d{4}/g;

const SSN_REGEX = /\b\d{3}[-\s]?\d{2}[-\s]?\d{4}\b/g;

const CREDIT_CARD_REGEX = /\b(?:\d{4}[-\s]?){3}\d{4}\b/g;

const ADDRESS_REGEX = /\b\d{1,5}\s+(?:[A-Za-z]+\s+){1,4}(?:St(?:reet)?|Ave(?:nue)?|Blvd|Boulevard|Rd|Road|Dr(?:ive)?|Ln|Lane|Way|Ct|Court|Pl|Place|Cir(?:cle)?|Hwy|Highway)\b\.?(?:\s*,?\s*(?:Apt|Suite|Unit|#)\s*\d+)?/gi;

const IP_ADDRESS_REGEX = /\b(?:\d{1,3}\.){3}\d{1,3}\b/g;

export interface RedactorConfig {
  maskChar?: string;
  preserveLength?: boolean;
  redactEmails?: boolean;
  redactPhones?: boolean;
  redactSSN?: boolean;
  redactCreditCards?: boolean;
  redactAddresses?: boolean;
  redactIPs?: boolean;
}

const DEFAULT_CONFIG: Required<RedactorConfig> = {
  maskChar: '*',
  preserveLength: false,
  redactEmails: true,
  redactPhones: true,
  redactSSN: true,
  redactCreditCards: true,
  redactAddresses: true,
  redactIPs: true,
};

function createMask(original: string, maskChar: string, preserveLength: boolean, label: string): string {
  if (preserveLength) {
    return maskChar.repeat(original.length);
  }
  return `[${label}]`;
}

function maskEmail(email: string, maskChar: string, preserveLength: boolean): string {
  if (preserveLength) {
    const [local, domain] = email.split('@');
    const maskedLocal = local[0] + maskChar.repeat(local.length - 1);
    const domainParts = domain.split('.');
    const maskedDomain = domainParts.map((part, i) => 
      i === domainParts.length - 1 ? part : maskChar.repeat(part.length)
    ).join('.');
    return `${maskedLocal}@${maskedDomain}`;
  }
  return '[EMAIL]';
}

function maskPhone(phone: string, maskChar: string, preserveLength: boolean): string {
  if (preserveLength) {
    return phone.replace(/\d/g, maskChar);
  }
  return '[PHONE]';
}

export function redact(text: string, config: RedactorConfig = {}): string {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  let result = text;

  if (cfg.redactEmails) {
    result = result.replace(EMAIL_REGEX, (match) => 
      maskEmail(match, cfg.maskChar, cfg.preserveLength)
    );
  }

  if (cfg.redactCreditCards) {
    result = result.replace(CREDIT_CARD_REGEX, (match) => 
      createMask(match, cfg.maskChar, cfg.preserveLength, 'CC')
    );
  }

  if (cfg.redactSSN) {
    result = result.replace(SSN_REGEX, (match) => 
      createMask(match, cfg.maskChar, cfg.preserveLength, 'SSN')
    );
  }

  if (cfg.redactPhones) {
    result = result.replace(PHONE_REGEX, (match) => 
      maskPhone(match, cfg.maskChar, cfg.preserveLength)
    );
  }

  if (cfg.redactAddresses) {
    result = result.replace(ADDRESS_REGEX, (match) => 
      createMask(match, cfg.maskChar, cfg.preserveLength, 'ADDRESS')
    );
  }

  if (cfg.redactIPs) {
    result = result.replace(IP_ADDRESS_REGEX, (match) => {
      if (match === '127.0.0.1' || match === '0.0.0.0') {
        return match;
      }
      return createMask(match, cfg.maskChar, cfg.preserveLength, 'IP');
    });
  }

  return result;
}

export function redactObject<T>(obj: T, config: RedactorConfig = {}): T {
  if (obj === null || obj === undefined) {
    return obj;
  }

  if (typeof obj === 'string') {
    return redact(obj, config) as T;
  }

  if (Array.isArray(obj)) {
    return obj.map(item => redactObject(item, config)) as T;
  }

  if (typeof obj === 'object') {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj)) {
      result[key] = redactObject(value, config);
    }
    return result as T;
  }

  return obj;
}

export function createRedactor(config: RedactorConfig = {}) {
  const cfg = { ...DEFAULT_CONFIG, ...config };
  
  return {
    redact: (text: string) => redact(text, cfg),
    redactObject: <T>(obj: T) => redactObject(obj, cfg),
  };
}

export const piiRedactor = createRedactor();
